File: apps/api/app/schemas/sessions.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field, ValidationInfo, field_validator

from app.models.enums import (
    InterviewDirection,
    InterviewDirectionV32,
    InterviewDurationV32,
    InterviewSessionStatus,
    InterviewStyle,
    InterviewStyleV32,
)
from app.models.legacy_mapping import (
    upgrade_legacy_direction,
    upgrade_legacy_duration,
    upgrade_legacy_style,
)
from app.schemas.common import PaginatedResponse, SchemaModel, TimestampedResponse
from app.schemas.frameworks import DirectionFramework


class InterviewConfigRequest(SchemaModel):
    """v3.2+ InterviewConfig request shape (F-307).

    Backward compat (L0): legacy v3.1 fields `direction` (singular) and
    `style` values like `standard_professional` continue to validate.
    Validators run in `mode='before'` to project the legacy palette
    onto the v3.2 Literal palette before strict typing kicks in.
    Any 1–3-direction list of v3.2 ids is accepted; a missing /
    empty `directions` falls back to the legacy `direction` field.
    """

    style: InterviewStyleV32
    # NB: `direction` is declared BEFORE `directions` so the
    # before-validator on `directions` can read the legacy field via
    # `info.data["direction"]` (Pydantic v2 only exposes already-
    # validated fields in `info.data`).
    direction: InterviewDirection | None = None
    # v3.2+ multi-select. Default to `[]` so v3.1 clients that omit the
    # field altogether (and only send `direction`) still hit the
    # before-validator that promotes the legacy single-direction. The
    # min_length=1 guard then fires AFTER the back-fill, so a payload
    # with neither `directions` nor `direction` still gets rejected.
    directions: list[InterviewDirectionV32] = Field(
        default_factory=list, min_length=1, max_length=3
    )
    duration_minutes: InterviewDurationV32

    @field_validator("style", mode="before")
    @classmethod
    def _upgrade_style(cls, v: Any) -> Any:
        return upgrade_legacy_style(v)

    @field_validator("duration_minutes", mode="before")
    @classmethod
    def _upgrade_duration(cls, v: Any) -> Any:
        return upgrade_legacy_duration(v)

    @field_validator("directions", mode="before")
    @classmethod
    def _fill_directions_from_legacy(
        cls, v: Any, info: ValidationInfo
    ) -> Any:
        # Old front-ends that send only `direction` (singular) get
        # auto-promoted to a one-element `directions` list.
        if (not v) and info.data.get("direction"):
            mapped = upgrade_legacy_direction(info.data["direction"])
            if mapped is not None:
                return [mapped]
        return v


class InterviewConfigResponse(TimestampedResponse):
    interview_session_id: UUID
    style: InterviewStyleV32
    directions: list[InterviewDirectionV32] = Field(default_factory=list)
    duration_minutes: InterviewDurationV32
    # `direction` echoes back whatever was stored in the legacy column
    # (string(64), nullable=False). The create_session shim now writes
    # the FIRST v3.2 direction id (e.g. "ai-insight") for v3.2+ clients
    # and falls back to the v3.1 enum value for legacy ones — so the
    # response type must accept either palette. Without this Union, GET
    # /sessions/{id} blows up with a Pydantic ValidationError on every
    # session whose primary direction is a v3.2 id.
    direction: InterviewDirectionV32 | InterviewDirection | None = None


class CreateSessionRequest(SchemaModel):
    asset_bundle_id: UUID
    config: InterviewConfigRequest


class SessionListRequest(SchemaModel):
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    status: InterviewSessionStatus | None = None


class CreateSessionResponse(SchemaModel):
    session_id: UUID
    status: InterviewSessionStatus
    direction_framework: DirectionFramework


class SessionSummary(TimestampedResponse):
    user_id: UUID
    candidate_asset_id: UUID
    status: InterviewSessionStatus
    started_at: datetime | None = None
    ended_at: datetime | None = None
    turn_count: int = Field(ge=0)
    config_snapshot: dict
    # M4-late additions for HistoryPage SessionTable. Both are optional
    # so v3.1 clients (and not-yet-graded sessions) keep working with
    # `None` / `[]` and the desktop renders the existing "—" placeholder.
    # Sourced from the *latest* InterviewReport row joined per session.
    latest_overall_score: int | None = None
    latest_weaknesses: list[str] = Field(default_factory=list, max_length=2)


class SessionDetailResponse(SessionSummary):
    config: InterviewConfigResponse | None = None
    direction_framework: DirectionFramework | None = None


class SessionListResponse(PaginatedResponse[SessionSummary]):
    pass


class EndSessionResponse(SchemaModel):
    session_id: UUID
    status: InterviewSessionStatus
    ended_at: datetime

File: apps/desktop/src/core/schemas/sessions.ts
import { z } from "zod";
import { TimestampedResponseSchema, paginatedResponseSchema } from "./common";
import {
  InterviewStyleV32Schema,
  InterviewDirectionV32Schema,
  InterviewDirectionSchema,
  InterviewDurationV32Schema,
  DirectionFrameworkSchema,
} from "./frameworks";

// ===== sessions.py enum =====

/**
 * Mirror of Python InterviewSessionStatus (StrEnum).
 * 13 values — L0 lock: deletion forbidden.
 */
export const InterviewSessionStatusSchema = z.enum([
  "created",
  "session_started",
  "turn_recording",
  "turn_transcribing",
  "turn_evaluating",
  "turn_compressing",
  "next_question_ready",
  "paused",
  "ended",
  "exited_early",
  "report_generating",
  "report_ready",
  "failed",
]);

export type InterviewSessionStatus = z.infer<
  typeof InterviewSessionStatusSchema
>;

// ===== sessions.py schemas (9 models) =====

/**
 * Mirror of Python InterviewConfigRequest(SchemaModel).
 * ★ L0-4 LOCK ★: InterviewConfig 4+6+4 三轴锁.
 *   style: InterviewStyleV32 — 4 values locked (frameworks.ts).
 *   directions: list[InterviewDirectionV32] — 6 values, min_length=1, max_length=3.
 *   duration_minutes: InterviewDurationV32 — 4 values locked (15/30/45/60, frameworks.ts).
 *   direction: legacy v3.1 single InterviewDirection — optional, 3 values.
 * Note: field_validators for legacy promotion omitted per §E1 (not in LangGraph.js scope).
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const InterviewConfigRequestSchema = z
  .object({
    style: InterviewStyleV32Schema, // ★ L0-4: 4 values locked
    direction: InterviewDirectionSchema.nullable().optional(), // legacy v3.1 single
    directions: z.array(InterviewDirectionV32Schema).min(1).max(3).default([]), // ★ L0-4: 6 values, 1-3 length
    duration_minutes: InterviewDurationV32Schema, // ★ L0-4: 4 values locked (15/30/45/60)
  })
  .strict();

export type InterviewConfigRequest = z.infer<
  typeof InterviewConfigRequestSchema
>;

/**
 * Mirror of Python InterviewConfigResponse(TimestampedResponse).
 * direction: InterviewDirectionV32 | InterviewDirection | None (union v3.2+v3.1).
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const InterviewConfigResponseSchema = TimestampedResponseSchema.extend({
  interview_session_id: z.string().uuid(),
  style: InterviewStyleV32Schema,
  directions: z.array(InterviewDirectionV32Schema).default([]),
  duration_minutes: InterviewDurationV32Schema,
  direction: z
    .union([InterviewDirectionV32Schema, InterviewDirectionSchema])
    .nullable()
    .optional(),
}).strict();

export type InterviewConfigResponse = z.infer<
  typeof InterviewConfigResponseSchema
>;

/**
 * Mirror of Python CreateSessionRequest(SchemaModel).
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const CreateSessionRequestSchema = z
  .object({
    asset_bundle_id: z.string().uuid(),
    config: InterviewConfigRequestSchema,
  })
  .strict();

export type CreateSessionRequest = z.infer<typeof CreateSessionRequestSchema>;

/**
 * Mirror of Python SessionListRequest(SchemaModel).
 * page: ge=1 default=1; page_size: ge=1 le=100 default=20.
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const SessionListRequestSchema = z
  .object({
    page: z.number().int().min(1).default(1),
    page_size: z.number().int().min(1).max(100).default(20),
    status: InterviewSessionStatusSchema.nullable().optional(),
  })
  .strict();

export type SessionListRequest = z.infer<typeof SessionListRequestSchema>;

/**
 * Mirror of Python CreateSessionResponse(SchemaModel).
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const CreateSessionResponseSchema = z
  .object({
    session_id: z.string().uuid(),
    status: InterviewSessionStatusSchema,
    direction_framework: DirectionFrameworkSchema,
  })
  .strict();

export type CreateSessionResponse = z.infer<typeof CreateSessionResponseSchema>;

/**
 * Mirror of Python SessionSummary(TimestampedResponse).
 * latest_weaknesses: max 2 items (per Pydantic max_length=2).
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const SessionSummarySchema = TimestampedResponseSchema.extend({
  user_id: z.string().uuid(),
  candidate_asset_id: z.string().uuid(),
  status: InterviewSessionStatusSchema,
  started_at: z.string().datetime().nullable().optional(),
  ended_at: z.string().datetime().nullable().optional(),
  turn_count: z.number().int().min(0),
  config_snapshot: z.record(z.unknown()),
  latest_overall_score: z.number().int().min(0).max(100).nullable().optional(),
  latest_weaknesses: z.array(z.string()).max(2).default([]),
}).strict();

export type SessionSummary = z.infer<typeof SessionSummarySchema>;

/**
 * Mirror of Python SessionDetailResponse(SessionSummary).
 * Extends SessionSummary with config and direction_framework fields.
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const SessionDetailResponseSchema = SessionSummarySchema.extend({
  config: InterviewConfigResponseSchema.nullable().optional(),
  direction_framework: DirectionFrameworkSchema.nullable().optional(),
}).strict();

export type SessionDetailResponse = z.infer<typeof SessionDetailResponseSchema>;

/**
 * Mirror of Python SessionListResponse(PaginatedResponse[SessionSummary]).
 * Uses paginatedResponseSchema factory from common.ts.
 */
export const SessionListResponseSchema = paginatedResponseSchema(
  SessionSummarySchema,
);

export type SessionListResponse = z.infer<typeof SessionListResponseSchema>;

/**
 * Mirror of Python EndSessionResponse(SchemaModel).
 * .strict() enforces no extra fields (v3.4 spec line 1417).
 */
export const EndSessionResponseSchema = z
  .object({
    session_id: z.string().uuid(),
    status: InterviewSessionStatusSchema,
    ended_at: z.string().datetime(),
  })
  .strict();

export type EndSessionResponse = z.infer<typeof EndSessionResponseSchema>;
